libvast/src/system/explore_command.cpp

```diff
@@ -0,0 +1,160 @@
+#include <caf/stateful_actor.hpp>
+
+#include <chrono>
+#include <csignal>
+
+using namespace caf;
+using namespace std::chrono_literals;
+
+namespace vast::system {
+
+caf::message
+explore_command(const command::invocation& invocation, caf::actor_system& sys) {
+  VAST_DEBUG_ANON(invocation);
+  const auto& options = invocation.options;
+  // Get query.
+  if (invocation.arguments.size() != 1) {
+    return caf::make_message(make_error(ec::invalid_query,
+                                        "expected query as first and only "
+                                        "positional argument"));
+  }
+  auto query = invocation.arguments.at(0);
+  auto before = caf::get_or(options, "explore.before", vast::duration{0s});
+  auto after = caf::get_or(options, "explore.after", vast::duration{0s});
+  // TODO: Add some way for users to pass "infinity" for either value.
+  if (before == 0s && after == 0s)
+    return caf::make_message(make_error(ec::invalid_configuration,
+                                        "Either the '--before' or the "
+                                        "'--after' option must "
+                                        "be present and non-zero."));
+  // Get a local actor to interact with `sys`.
+  caf::scoped_actor self{sys};
+  // TODO: Add --format option to select output format.
+  auto out = detail::make_output_stream(
+    std::string{format::json::writer::defaults::write}, false);
+  if (!out)
+    return make_message(out.error());
+  caf::actor writer
+    = sys.spawn(sink<format::json::writer>,
+                format::json::writer{std::move(*out)}, max_events);
+  if (!writer) {
+    return make_message(ec::unspecified, "could not spawn writer");
+  }
+  auto writer_guard = caf::detail::make_scope_guard(
+    [&] { self->send_exit(writer, caf::exit_reason::user_shutdown); });
+  self->monitor(writer);
+  // Get VAST node.
+  auto node_opt
+    = system::spawn_or_connect_to_node(self, options, content(sys.config()));
+  if (auto err = caf::get_if<caf::error>(&node_opt))
+    return caf::make_message(std::move(*err));
+  auto& node = caf::holds_alternative<caf::actor>(node_opt)
+                 ? caf::get<caf::actor>(node_opt)
+                 : caf::get<scope_linked_actor>(node_opt).get();
+  VAST_ASSERT(node != nullptr);
+  // Start signal monitor.
+  std::thread sig_mon_thread;
+  auto guard = system::signal_monitor::run_guarded(
+    sig_mon_thread, sys, defaults::system::signal_monitoring_interval, self);
+  // Spawn exporter for the passed query
+  auto spawn_exporter
+    = command::invocation{invocation.options, "spawn exporter", {query}};
+  VAST_DEBUG(&invocation, "spawns exporter with parameters:", spawn_exporter);
+  auto exporter = spawn_at_node(self, node, spawn_exporter);
+  if (!exporter)
+    return caf::make_message(std::move(exporter.error()));
+  auto exp_guard = caf::detail::make_scope_guard(
+    [&] { self->send_exit(*exporter, caf::exit_reason::user_shutdown); });
```

Comment:
Can you add a debug or trace message in the scope guard?

Reply:
Also, this one may as well be renamed to `exporter_guard` to be explicit. I had to stop and think for a bit what `exp_guard` is.